linkcheck/url.py

```diff
@@ -258,12 +269,13 @@ def url_fix_wayback_query(path):
 
 def url_parse_query (query, encoding=None):
     """Parse and re-join the given CGI query."""
-    if isinstance(query, unicode):
+    if isinstance(query, str_text):
         if encoding is None:
             encoding = url_encoding
         query = query.encode(encoding, 'ignore')
     # if ? is in the query, split it off, seen at msdn.microsoft.com
     append = ""
+    query = decode_for_unquote(query)
```

Comment:
This feels wrong.  We've only just encoded it to `encoding or url_encoding` (which is not necessarily UTF-8) three lines up from here, and now we're decoding as UTF-8.

Reply:
I tried taking it and got 67 failures including:
```
_____________________________________________ TestMailBad.test_error_mail ______________________________________________
[gw14] linux -- Python 3.6.8 /usr/bin/python3.6

self = <tests.checker.test_mail_bad.TestMailBad testMethod=test_error_mail>

    def test_error_mail (self):
        # too long or too short
>       self.mail_error(u"mailto:@")

tests/checker/test_mail_bad.py:28: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/checker/__init__.py:279: in mail_error
    return self.mail_test(addr, u"error", **kwargs)
tests/checker/__init__.py:283: in mail_test
    url = self.norm(addr)
tests/checker/__init__.py:191: in norm
    return linkcheck.url.url_norm(url, encoding=encoding)[0]
linkcheck/url.py:334: in url_norm
    urlparts[3] = url_parse_query(urlparts[3], encoding=encoding)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

query = b'', encoding = None

    def url_parse_query (query, encoding=None):
        """Parse and re-join the given CGI query."""
        if isinstance(query, str_text):
            if encoding is None:
                encoding = url_encoding
            query = query.encode(encoding, 'ignore')
        # if ? is in the query, split it off, seen at msdn.microsoft.com
        append = ""
>       while '?' in query:
E       TypeError: a bytes-like object is required, not 'str'

linkcheck/url.py:278: TypeError
```

Reply:
On the other hand removing the section above:

```diff
--- a/linkcheck/url.py
+++ b/linkcheck/url.py
@@ -269,10 +269,6 @@ def url_fix_wayback_query(path):
 
 def url_parse_query (query, encoding=None):
     """Parse and re-join the given CGI query."""
-    if isinstance(query, str_text):
-        if encoding is None:
-            encoding = url_encoding
-        query = query.encode(encoding, 'ignore')
     # if ? is in the query, split it off, seen at msdn.microsoft.com
     append = ""
     query = decode_for_unquote(query)
```

and the python3 branch passes tests for 2.7 and 3.6.
Although there is a warning with 2.7:
```
linkcheck/checker/ftpurl.py:140
  /var/tmp/portage/net-analyzer/linkchecker-9999-r100/work/linkchecker-9999/linkcheck/checker/ftpurl.py:140: UnicodeWarning: Unicode equal comparison failed to convert both arguments to Unicode - interpreting them as being unequal
    if self.filename in files:
```

I'm tempted to ignore that though and just press on bearing in mind we will go Python 3 only at the end of this.

Reply:
> Although there is a warning with 2.7:
> 
> ```
> linkcheck/checker/ftpurl.py:140
>   /var/tmp/portage/net-analyzer/linkchecker-9999-r100/work/linkchecker-9999/linkcheck/checker/ftpurl.py:140: UnicodeWarning: Unicode equal comparison failed to convert both arguments to Unicode - interpreting them as being unequal
>     if self.filename in files:
> ```

Actually looks like that is nothing to do with this, but a consequence of enablin the FTP test.

Reply:
> I'm tempted to ignore that though and just press on bearing in mind we will go Python 3 only at the end of this.

I'm tempted to do that as well since I cannot muster the energy to analyze all the sources to figure out what's the right thing to do....